ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -253,56 +253,62 @@ public boolean updateAccounts(Collection<Account> accounts) {
+    for (Container existingContainer : account.getAllContainers()) {
+      for (Container newContainer : containers) {
+        // make sure there is no conflict container (conflict means a container with same name but different attributes already exists).
+        if (existingContainer.getName().equals(newContainer.getName())) {
+          if (existingContainer.isSameContainer(newContainer)) {
+            // If an exactly same container already exists, we directly return its id. Adding same container multiple times
+            // should be no-op.
+            existingContainers.add(existingContainer);
+          } else {
+            throw new IllegalArgumentException("There is a conflicting container in account " + accountName);
+          }
         }
       }
     }
-    // if code reaches here, it means no duplicate container in this account
-    // generate container id for new container
+    newContainers.removeAll(existingContainers);
+
+    // if code reaches here, it means no conflicting container in this account
     short nextContainerId = account.getAllContainers()
         .stream()
         .map(Container::getId)
         .max(Short::compareTo)
         .map(maxId -> (short) (maxId + 1))
         .orElse(config.containerIdStartNumber);
     // construct a container based on input container and next containerId
-    Container containerToAdd =
-        new ContainerBuilder(newContainer).setId(nextContainerId).setParentAccountId(account.getId()).build();
-    account.updateContainerMap(containerToAdd);
-    List<Account> accounts = Collections.singletonList(account);
-    boolean hasSucceeded = false;
-    if (accountInfoMapRef.get().hasConflictingAccount(accounts)) {
-      logger.debug("Accounts={} conflict with the accounts in local cache. Cancel the update operation.", accounts);
-      accountServiceMetrics.updateAccountErrorCount.inc();
-    } else {
-      hasSucceeded = accountMetadataStore.updateAccounts(accounts);
+    List<Container> createdContainers = new ArrayList<>();
+    for (Container container : newContainers) {
+      createdContainers.add(
+          new ContainerBuilder(container).setId(nextContainerId).setParentAccountId(account.getId()).build());
+      ++nextContainerId;
     }
+    // In case updating account metadata store failed, we do a deep copy of original account. Thus, we don't have to
+    // revert changes in original account when there is a failure.
+    Account accountCopy = new AccountBuilder(account).build();
+    accountCopy.updateContainerMap(createdContainers);
+    boolean hasSucceeded = updateAccounts(Collections.singletonList(accountCopy));
     if (!hasSucceeded) {
       throw new IllegalStateException("Account update failed for " + accountName);
     }
-    return new Pair<>(accountId, String.valueOf(nextContainerId));
+    // after metadata store is successfully updated, we safely update original account
+    account.updateContainerMap(createdContainers);
```

Comment:
Even if we don't update in-memory account here, may be it might eventually get updated later when fetchAndUpdateCache() is called by background updater thread or on ZK notification (similar to how accounts in cache get updated today after updateAccounts() is called from PostAccountsHandler).

Reply:
Yes, that's true. However, I think updating the in-mem account immediately doesn't do harm to the correctness. It ensures the in-mem account immediately has latest view (in case ZK notification missed and there is subsequent request against new container immediately after the creation).